### PR TITLE
applications: serial_lte_modem: add GPIO AT command

### PR DESCRIPTION
**Pull request by pirun**
The GPIO AT commands support configure/read/write/toggle GPIO.
Currently the supported commands are:

AT#XGPIOCFG: Configure GPIO pins as input pin, output pin or custom functions.
AT#XGPIO: Access GPIO pins.

Signed-off-by: Lee <pirun.lee@nordicsemi.no>

### CODE REVIEW

File: applications/serial_lte_modem/prj.conf [hunk 1]
```diff
@@ -102,3 +102,4 @@ CONFIG_UART_0_NRF_HW_ASYNC_TIMER=2
 CONFIG_SLM_AGPS=n
 CONFIG_SLM_PGPS=n
 CONFIG_SLM_CELL_POS=n
+CONFIG_SLM_GPIO=n
```
Comment (junqingzou) [inline applications/serial_lte_modem/prj.conf:105]: This is redundant

File: applications/serial_lte_modem/src/gpio/Kconfig [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+
+config SLM_GPIO
+	bool "GPIO support in SLM"
+	default n
+	select GPIO
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/Kconfig:9]: Default should be 'y' as required by PMT
CI team does not want to see "select", change to "depends on" and select in prj.conf
BTW, CONFIG_GPIO has already been configured in prj.conf.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include "slm_util.h"
+#include "slm_at_gpio.h"
+
+LOG_MODULE_REGISTER(slm_gpio, CONFIG_SLM_LOG_LEVEL);
+
+/* global functions defined in different resources */
+void rsp_send(const uint8_t *str, size_t len);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:16]: Please include "slm_at_host.h" instead, using the APIs defined there

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+/* global variable defined in different resources */
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
+
+const struct device *gpio_dev;
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:22]: static

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+/* global variable defined in different files */
+extern struct k_work_q slm_work_q;
+
+struct slm_gpio_pin_t {
+	sys_snode_t node;
+	gpio_pin_t pin;
+	uint16_t op;
+};
+
+gpio_flags_t convert_flags(uint16_t op)
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:34]: static

File: doc/nrf/releases/release-notes-changelog.rst [hunk 6]
```diff
@@ -70,7 +70,7 @@ nRF9160
     * Added IPv6 support to all SLM services.
     * Added the GNSS service to replace the existing GPS test functionality.
     * Added the optional support of location services from nRF Cloud, like A-GPS, P-GPS, and cellular positioning.
-
+    * Added GPIO AT commands.
```
Comment (junqingzou) [inline doc/nrf/releases/release-notes-changelog.rst:73]: Please don't change this file and wait for NCSv1.7.0 first

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.h [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+#define SLM_GPIOC_OP_IN_PU	21	/* Enables pin as input. Use internal pull up resistor. */
+#define SLM_GPIOC_OP_IN_PD	22	/* Enables pin as input. Use internal pull down resistor. */
+
+/**@brief GPIO operations. */
+enum slm_gpio_operations {
+	SLM_GPIO_OP_WRITE,
+	SLM_GPIO_OP_READ,
+	SLM_GPIO_OP_TOGGLE
+};
+
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.h:32]: Definitions and enum should be in .c file instead

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.h [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+ */
+int slm_at_gpio_uninit(void);
+
+/**
+ * @brief GPIO configuration.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int do_gpio_pin_configure_set(uint16_t op, gpio_pin_t pin);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.h:55]: this is an internal function and not an API

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+		break;
+	default:
+		LOG_ERR("Fail to convert gpio flag");
+		break;
+	}
+
+	return gpio_flags;
+}
+
+int do_gpio_pin_configure_set(uint16_t op, gpio_pin_t pin)
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:59]: static

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+{
+	int err = 0;
+	gpio_flags_t gpio_flags = 0;
+	struct slm_gpio_pin_t *slm_gpio_pin = NULL, *cur = NULL, *next = NULL;
+	gpio_port_pins_t pin_mask = 0;
+
+	LOG_INF("op:%hu pin:%hu", op, pin);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:66]: LOG_DBG?

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+	/* Convert SLM GPIO flag to zephyr gpio pin configuration flag */
+	gpio_flags = convert_flags(op);
+	if (gpio_flags == UINT32_MAX) {
+		LOG_ERR("Fail to configure pin.");
+		return -EINVAL;
+	}
+
+	/* Trace gpio list */
+	if (sys_slist_peek_head(&slm_gpios) != NULL) {
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&slm_gpios, cur,
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:83]: Put it in one line, max 100 chars allowed

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+			return -ENOBUFS;
+		}
+		memset(slm_gpio_pin, 0, sizeof(struct slm_gpio_pin_t));
+		sys_slist_append(&slm_gpios, &slm_gpio_pin->node);
+	}
+
+	err = gpio_pin_configure(gpio_dev, pin, gpio_flags);
+	if (err) {
+		LOG_ERR("GPIO_0 config error: %d", err);
+		//TODO: free and remove node
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:104]: Comment by // is not allowed in NCS, use /* */
Also what we really need to do here?

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 13]
```diff
@@ -0,0 +1,312 @@
...
+			return err;
+		}
+		/* Remove callback */
+		pin_mask &= ~BIT(pin);
+	}
+
+	return err;
+}
+
+int do_gpio_pin_configure_read(void)
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:124]: static

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 14]
```diff
@@ -0,0 +1,312 @@
...
+				sprintf(rsp_buf, "%hu,%hu\r\n", cur->op, cur->pin);
+				rsp_send(rsp_buf, strlen(rsp_buf));
+			}
+		}
+	}
+
+	return err;
+}
+
+int do_gpio_pin_operate(uint16_t op, gpio_pin_t pin, uint16_t value)
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:146]: static

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+{
+	int ret = 0;
+	struct slm_gpio_pin_t *cur = NULL, *next = NULL;
+
+	if (sys_slist_peek_head(&slm_gpios) != NULL) {
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&slm_gpios, cur,
+					     next, node) {
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:153]: Use one line

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+		}
+	}
+
+	return 0;
+}
+
+/**@brief handle AT#XGPIOCFG commands
+ *  AT#XGPIOCFG=<op>,<pin>
+ *  AT#XGPIOCFG?
+ *  AT#XGPIOCFG=?
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:192]: Not supported right?

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 17]
```diff
@@ -0,0 +1,312 @@
...
+			return err;
+		}
+		err = do_gpio_pin_configure_set(op, (gpio_pin_t)pin);
+		break;
+	case AT_CMD_TYPE_READ_COMMAND:
+		err = do_gpio_pin_configure_read();
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
+		break;
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:221]: Should support or, if not, remove this case

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 18]
```diff
@@ -0,0 +1,312 @@
...
+ *  AT#XGPIOCFG=?
+ */
+int handle_at_gpio_configure(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	uint16_t pin = 0xff, op = 0xff;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) == 0) {
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:201]: No need this call, redundant.
at_param_ APIs does check index

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 19]
```diff
@@ -0,0 +1,312 @@
...
+			return -EINVAL;
+		}
+		err = at_params_short_get(&at_param_list, 1, &op);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:204]: I believe that it should be at_param_unsigned_short_get()

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 20]
```diff
@@ -0,0 +1,312 @@
...
+		if (err < 0) {
+			LOG_ERR("Fail to get op: %d", err);
+			return err;
+		}
+		err = at_params_short_get(&at_param_list, 2, &pin);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:209]: again, unsigned_short_get, declared uint16_t pin above

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 21]
```diff
@@ -0,0 +1,312 @@
...
+ *  AT#XGPIO=?
+ */
+int handle_at_gpio_operate(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+	uint16_t pin = 0xff, op = 0xff, value = 0xff;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) == 0) {
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:242]: Redundant

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+			return -EINVAL;
+		}
+		err = at_params_short_get(&at_param_list, 1, &op);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:245]: unsigned

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+		if (err < 0) {
+			LOG_ERR("Fail to get op: %d", err);
+			return err;
+		}		
+		if (op > SLM_GPIO_OP_TOGGLE) {
+			LOG_ERR("Fail to operate gpio: %d", op);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:251]: Invalid parameter

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+			return -EINVAL;
+		}
+		err = at_params_short_get(&at_param_list, 2, &pin);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:254]: unsigned

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 25]
```diff
@@ -0,0 +1,312 @@
...
+		break;
+
+	default:
+		break;
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XGPIOCFG commands
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:230]: #XGPIO

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 26]
```diff
@@ -0,0 +1,312 @@
...
+				if (err < 0) {
+					LOG_ERR("Fail to get value: %d", err);
+					return err;
+				}
+				if (value != 1 && value != 0) {
+					LOG_ERR("Fail to set gpio value: %d", value);
+					return -EINVAL;
+				}
+			}
+			else {
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:275]: No new line

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 27]
```diff
@@ -0,0 +1,312 @@
...
+			LOG_ERR("Fail to get pin: %d", err);
+			return err;
+		}
+		if (pin > MAX_GPIO_PIN) {
+			LOG_ERR("Incorrect <pin>: %d", pin);
+			return -EINVAL;
+		}
+		if (op == SLM_GPIO_OP_WRITE) {
+			if (at_params_valid_count_get(&at_param_list) == 4) {
+				err = at_params_short_get(&at_param_list, 3, &value);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:265]: 0 and 1 only, so unsigned

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 28]
```diff
@@ -0,0 +1,312 @@
...
+				if (err < 0) {
+					LOG_ERR("Fail to get value: %d", err);
+					return err;
+				}
+				if (value != 1 && value != 0) {
+					LOG_ERR("Fail to set gpio value: %d", value);
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:271]: Invalid parameter

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 29]
```diff
@@ -0,0 +1,312 @@
...
+				}
+			}
+			else {
+				LOG_ERR("Fail to get value: %d", value);
+				return -EINVAL;
+			}
+		}
+		err = do_gpio_pin_operate(op, (gpio_pin_t)pin, value);
+		break;
+	case AT_CMD_TYPE_READ_COMMAND:
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:282]: No support so remove this case

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 30]
```diff
@@ -0,0 +1,312 @@
...
+		break;
+
+	case AT_CMD_TYPE_TEST_COMMAND:
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:285]: Need implementation

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 31]
```diff
@@ -0,0 +1,312 @@
...
+
+	default:
+		break;
+	}
+	return err;
+}
+
+int slm_at_gpio_init(void)
+{
+	int err = 0;
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:296]: redundant

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 32]
```diff
@@ -0,0 +1,312 @@
...
+
+	gpio_dev = device_get_binding(DT_LABEL(DT_NODELABEL(gpio0)));
+	if (gpio_dev == NULL) {
+		LOG_ERR("GPIO_0 bind error");
+		return -EIO;
+	}
+
+	return err;
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:304]: return 0

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 33]
```diff
@@ -0,0 +1,312 @@
...
+}
+
+int slm_at_gpio_uninit(void)
+{
+	int err = 0;
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:309]: Plan to do something here?
Otherwise just return 0

File: applications/serial_lte_modem/src/gpio/Kconfig [hunk 34]
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright (c) 2020 Nordic Semiconductor ASA
+#
```
Comment (lats1980) [inline applications/serial_lte_modem/src/gpio/Kconfig:3]: 2021

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 12 above)
Comment (rlubos) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c:104]: I guess free the memory allocated with `k_malloc`.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 35]
```diff
@@ -0,0 +1,312 @@
...
+
+	if (op == SLM_GPIOC_OP_DISABLE) {
+		/* Disable interrupt */
+		err = gpio_pin_interrupt_configure(gpio_dev, pin, GPIO_INT_DISABLE);
+		if (err) {
+			LOG_ERR("Interface pin interrupt config error: %d", err);
+			return err;
+		}
+		/* Remove callback */
+		pin_mask &= ~BIT(pin);
```
Comment (rlubos) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:118]: Once you disable GPIO, shouldn't you free the allocated memory as well? Or is it such a design, that once you allocate a memory for a GPIO, it stays allocated till reboot? If so, perhaps it'd be better to pre-allocate the memory statically, and readuce the heap size?

**Review by rlubos — COMMENTED** [review]
A few extra comments

File: applications/serial_lte_modem/prj.conf (hunk 1 above)
Comment (pirun) [reply applications/serial_lte_modem/prj.conf]: Removed

File: applications/serial_lte_modem/src/gpio/Kconfig (hunk 2 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/Kconfig]: Modify default to y.
Change to depends on GPIO

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 3 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: include slm_at_host.h

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 4 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: add static to local variable

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 5 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Add static to local function

File: doc/nrf/releases/release-notes-changelog.rst (hunk 6 above)
Comment (pirun) [reply doc/nrf/releases/release-notes-changelog.rst]: Removed from release note.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.h (hunk 7 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.h]: Move enum to c file.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.h (hunk 8 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.h]: Agree, remove declaration.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 9 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Add static to local function.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 10 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change LOG_DBG

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 11 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Adjust to one line.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 12 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Modify logic here.
If configure gpio failed, don't add gpio node to list.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 13 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Add static to local function.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 14 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Add static to local function.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 15 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Adjust to one line.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 16 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change comments here.
Read support .
Test not support.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 17 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Remove placeholder.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 18 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Remove here, let at_params_get_ API check index.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 19 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change to at_params_unsigned_short_get.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 20 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change to at_params_unsigned_short_get.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 21 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Remove here, let at_params_get_ API check index.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 22 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change to at_params_unsigned_short_get.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 24 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change to at_params_unsigned_short_get.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 25 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Fix typo.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 29 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Remove placeholder here.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 30 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Remove placeholder.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 33 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: return 0

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 36]
```diff
@@ -0,0 +1,312 @@
...
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
+
+const struct device *gpio_dev;
+static sys_slist_t slm_gpios = SYS_SLIST_STATIC_INIT(&slm_gpios);
+
+/* global variable defined in different files */
+extern struct k_work_q slm_work_q;
+
+struct slm_gpio_pin_t {
```
[the remark this answers is not in the document]
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c]: Change sturcture name.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.h [hunk 37]
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef SLM_AT_GPIO_
+#define SLM_AT_GPIO_
+
+/**@file slm_at_GPIO.h
```
[the remark this answers is not in the document]
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.h]: Fix typo.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 35 above)
Comment (pirun) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c:132]: Remove slm_gpio_pin in sys_lists and free slm_gpio_pin.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c (hunk 23 above)
Comment (junqingzou) [reply applications/serial_lte_modem/src/gpio/slm_at_gpio.c:258]: Please consider a better error message

**Review by junqingzou — APPROVED** [review]
Nice work!

**Comment by pirun** [comment]
Revised as suggestion, please review again.

File: CODEOWNERS [hunk 38]
```diff
@@ -95,6 +95,7 @@ Kconfig*                                  @tejlmand
 /modules/                                 @tejlmand
 /modules/mcuboot/                         @nvlsianpu
 /modules/cjson/                           @simensrostad @plskeggs @sigvartmh
+/samples/                                 @nrfconnect/ncs-test-leads
```
Comment (rlubos) [inline CODEOWNERS:98]: I guess something went wrong with rebasing? This commit doesn't seem to belong to this PR.

**Review by rlubos — APPROVED** [review]
LGTM

**Review by rlubos — CHANGES_REQUESTED** [review]
Ah, drop the needless commit first, then LGTM.

**Review by rlubos — APPROVED** [review]

File: CODEOWNERS (hunk 38 above)
Comment (pirun) [reply CODEOWNERS]: Hi Robert,
I'm still trying to fix rebasing error.
Please revoke approval , thanks.

File: applications/serial_lte_modem/src/gpio/slm_at_gpio.c [hunk 39]
```diff
@@ -0,0 +1,305 @@
...
+#include <drivers/gpio.h>
+#include "slm_util.h"
+#include "slm_at_gpio.h"
+#include "slm_at_host.h"
+
+LOG_MODULE_REGISTER(slm_gpio, CONFIG_SLM_LOG_LEVEL);
+
+/* global variable defined in different resources */
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/slm_at_gpio.c:19]: Size is not correct, check other modules

File: applications/serial_lte_modem/src/gpio/Kconfig [hunk 40]
```diff
@@ -0,0 +1,9 @@
...
+# Copyright (c) 2021 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+
+config SLM_GPIO
```
Comment (junqingzou) [inline applications/serial_lte_modem/src/gpio/Kconfig:6]: New CONFIG item needs to be described in "slm_description.rst"

**Review by FrancescoSer — APPROVED** [review]

**Review by rlubos — COMMENTED** [review]
Please squash the doc review commit.

Did you remove the changelog entry by accident? There seemed to be the following line before:
```
 * Added GPIO AT commands.
```

**Review by lats1980 — APPROVED** [review]
LGTM.
